c10/cuda/CUDAFunctions.h

```diff
@@ -89,12 +97,18 @@ C10_CUDA_API void __inline__ memcpy_and_sync(
     (*interp)->trace_gpu_stream_synchronization(
         c10::kCUDA, reinterpret_cast<uintptr_t>(stream));
   }
+  if (width_in_bytes == -1) {
 #if defined(TORCH_HIP_VERSION) && (TORCH_HIP_VERSION >= 301)
-  C10_CUDA_CHECK(hipMemcpyWithStream(dst, src, nbytes, kind, stream));
+    C10_CUDA_CHECK(hipMemcpyWithStream(dst, src, nbytes, kind, stream));
 #else
-  C10_CUDA_CHECK(cudaMemcpyAsync(dst, src, nbytes, kind, stream));
-  C10_CUDA_CHECK(cudaStreamSynchronize(stream));
+    C10_CUDA_CHECK(cudaMemcpyAsync(dst, src, nbytes, kind, stream));
+    C10_CUDA_CHECK(cudaStreamSynchronize(stream));
 #endif
+  } else {
+    C10_CUDA_CHECK(cudaMemcpy2DAsync(
```

Comment:
cc @jeffdaily, this version of copy2d PR passes with ROCm, so probably the test errors in the previous one were legit, but if you have an API similar to hipMemcpyWithStream that's more efficient than explicit synchronization you should probably put it here

Reply:
I'm sure we don't have a hipMemcpy2DWithStream.

At the time when hipMemcpyWithStream was created we were looking for opportunities to positively differentiate ourselves from the CUDA API if we found missing capabilities or chances to optimize.  This was an early optimization that may have confused users more than any perf benefit it provided.  When we profile ROCm vs CUDA side by side we get questions why the profiles look different, such as "why is HIP calling blocking copy while CUDA's profile shows a non-blocking copy?" which misses the point that they're both effectively a blocking copy.  Chances are that our HIP runtime is just as efficient now making the two API calls back to back.